Add Result concept and exercise

ceddlyburge opened the pull request:
Fixes #427

jiegillet commented on exercises/concept/go/.docs/instructions.md:
```diff
@@ -0,0 +1,43 @@
+- You can not place a stone where it would capture itself (libertyRule)
+- You can not use the same point twice (koRule)
+
+## 1. Apply the rules
```
I think this exercise is a little bit too much in one go and could benefit from being split into 3 or 4 steps.

 Maybe the student could apply all these rules one at a time. Like `applyOneStonePerPointRule : Game -> (Game -> Result String Game) -> Result String Game ` and so on. Also, implementing `applyError : Game -> (Game -> Result String Game) -> Game` would be a nice extra step to practice pattern matching.

I also thing the `Rule` type alias might be too much abstraction. In an exercise concept about `Result`, it's OK to see `Result` everywhere. Maybe defining it just before the final `applyRules` would be OK.

I also think I would help to see a `board : Dict (Int, Int) (Maybe Player)` in `Game` or something just to make it look more like a game of Go?

ceddlyburge replied:
It is quite a chunky exercise, but you end up with something that looks like good real world code, and requires the use of all the common Result functions / syntax. If you have thought of a nice way of splitting it in to small steps while keeping all this please suggest a change / pr for the example code.

I don't think I understand applying the rules once at a time. That is basically just calling the function I think?

We could remove the rule abstraction, but the type annotations would become very big, and adding a type alias is something you would do in production code, so I think its ok. Maybe we could add a comment or something to the instructions to explain it in more detail (although `type alias` is a requirement of this exercise I think)

Adding a `board` isn't required for the exercise, so potentially makes things more complex, but maybe we could add it with a comment, or add it as a comment?

jiegillet replied:
I'll see if I can come up with something that works...

jiegillet reviewed (commented):
Hello, I have been interested in the Elm track for a while, so I thought I would help out a bit. I hope that it's OK that I drop in with a review unannounced like this.

Anyway, awesome job! Writing a concept exercise is a ton of work, I appreciate that!

I left a few comments with small changes in the text, I'm happy to discuss any of them. I also think that the exercise itself should be expanded out a bit, for the sake of easing in beginners, and I left some suggestions about that as well.

ceddlyburge commented on concepts/result/about.md:
```diff
@@ -0,0 +1,113 @@
+    --> "That is not a number!"
+
+howOldAmI "32"
+    --> "You are 32"
```
I think the examples already look different from function definitions (you couldn't have "32" in a function definition). This is also the way that all the other example code is written, so I think we should leave it as it is.

jiegillet replied:
It tripped me up a bit, because in Haskell it would be a valid way to start a definition. Also in other languages like Elixir, so it might be confusing for other people than me.
Then maybe adding a type signature?
```elm
bananasAge : String
bananasAge = howOldAmI "Bananas"
```
If not, I might do a PR to consistently change all the example code :)

ceddlyburge commented on exercises/concept/go/src/Go.elm:
```diff
@@ -0,0 +1,40 @@
+
+applyRules : Game -> Rule -> NonValidatingRule -> Rule -> Rule -> Game
+applyRules game oneStonePerPointRule captureRule libertyRule koRule =
+    Debug.todo "Please implement this function"
```
I think this is fine, and that Debug gets imported automatically: https://dennisreimann.de/articles/elm-modules-import.html

jiegillet replied:
Neat, I didn't know that!

ceddlyburge commented on concepts/result/introduction.md:
```diff
@@ -0,0 +1,16 @@
+# Introduction
+
+The `Result` type is the solution in the Elm language for error reporting and an enabler for error recovery. It is somewhat similar to the `Maybe` type, but where Maybe can be returned by functions that may fail, it doesn't tell you anything about why it failed. Result includes this context about why something failed.
```
I'm not a huge fan of the about.md and introudction.md stuff, being as they are similar and repeated. I think the way it is here is the way that exercism recommend though, with the concept introduction being a brief explainer of the topic that is shown early on in the process, and the other things shown when the student is thinking about the exercise or referring back to it.

jiegillet replied:
Sorry, that was unclear, I meant make the same textual changes you made to `about.md`. Or maybe that should be done once `about.md` is finalized.

I share your view though, that intro system is clunky.

ceddlyburge commented on exercises/concept/go/.docs/introduction.md:
```diff
@@ -0,0 +1,113 @@
+# Introduction
```
I'm not a huge fan of the about.md and introudction.md stuff, being as they are similar and repeated. I think the way it is here is the way that exercism recommend though, with the concept introduction being a brief explainer of the topic that is shown early on in the process, and the other things shown when the student is thinking about the exercise or referring back to it.

ceddlyburge commented on concepts/result/about.md:
````diff
@@ -0,0 +1,113 @@
+```
+
+The vertical bar `|` in the type definition of `Result` means "OR".
+It indicates that it can either be `Ok` something OR be `Error` something.
````
I don't quite understand this comment, could you add some more details?

jiegillet replied:
Lines 38 and 39 are a direct comment on the type definition on line 10, so they should be moved up. 

Additionally, I felt that those two lines were a bit vague so I proposed an alternative phrasing.

ceddlyburge replied:
Ok, I've made that change.

ceddlyburge commented:
Hi Jie, thanks for the review! I've committed most of those changes.

If you are keen to be more invovled then take a look at the issues and send in a PR. Concept exercises are probably the thing we are focussing on most at the moment.
https://github.com/exercism/elm/issues

Cheers, Cedd

jiegillet reviewed (commented):
I replied to a couple of points...

I am keen on getting a bit more involved, I've actually started to work on a [concept for an Elm Analyzer](https://github.com/exercism/elm-analyzer/pull/12), which go hand in hand with concept exercises (for example, please do open an issue on that repo to check that `andThen` and `map` are actually used instead of pattern matching for example).

I'll consider creating a concept exercise. I've done [one before for the Elixir track](https://github.com/exercism/elixir/tree/main/exercises/concept/new-passport), it was quite a bit of work. Now that I think about it though, it's an exercise very suitable for this concept as well ^^

ceddlyburge commented on concepts/result/about.md:
````diff
@@ -0,0 +1,113 @@
+```
+
+This is known as a "custom type" definition in Elm terminology.
+Custom types are introduced in details in another concept of this track but we won't need it to understand specifically `Result`.
````
I am going to leave this for now, and we can revisit it later. There aren't that many concept exercises yet so I think it is too early to make good decisions about how the dependency tree for the concepts should look.

ceddlyburge commented:
Creating an analyser should be a fun task, and will be helpful. It is quite a lot of work though, as each exercise will require its own rules.

I have created a few concept exercises now, and its definitely a lot easier when you can port one from another language! Especially from another functional language or one with similar concepts!

ceddlyburge commented:
Hi @exercism/reviewers, Matthieu is super busy, and its only me and him maintaining the Elm track at the moment, so there isn't anybody to review / approve my pull requests at the moment. I emailed the only other maintainer, but he has been inactive for years, and didn't reply.
Could somebody take a look?
Thanks, Cedd

ErikSchierboom commented on exercises/concept/go/.docs/introduction.md:
```diff
@@ -0,0 +1,113 @@
+# Introduction
```
It looks like the contents of this file match the concept's `about.md`, but it should be much closer to the `introduction.md`. As per [the docs](https://exercism.org/docs/building/tracks/concept-exercises#h-file-docs-introduction-md), the exercise's `introduction.md` file should give the student just enough context to figure out the solution themselves.

ceddlyburge replied:
Good idea, I've done this. The interplay of these 3 files is a source of confusion for me, it feels like it might be possible to find a better way to structure the information, although I realise there are reasons ...

ErikSchierboom replied:
There are indeed reasons :) We do offer the option of using `configlet generate`: https://exercism.org/docs/building/configlet/generating-documents This allows you to basically copy the contents of the concept introduction to the exercise introduction without repeating it.

ErikSchierboom commented on exercises/concept/go/.meta/Exemplar.elm:
```diff
@@ -0,0 +1,53 @@
+                |> Result.andThen koRule
+                |> Result.map changePlayer
+    in
+    case updatedGameOrError of
```
Does the student know how to do pattern matching at this point?

ceddlyburge replied:
Yep, this is covered in the Custom Types concept

ErikSchierboom commented on exercises/concept/go/src/Go.elm:
```diff
@@ -0,0 +1,40 @@
+module Go exposing (..)
+
+
+type Player
+    = Black
+    | White
+
+
+type alias Game =
+    { whiteCapturedStones : Int
+    , blackCapturedStones : Int
+    , player : Player
+    , error : String
+    }
+
+
+type alias Rule =
+    Game -> Result String Game
+
+
+type alias NonValidatingRule =
+    Game -> Game
+
+
+changePlayer : Game -> Game
+changePlayer game =
+    let
+        newPlayer =
+            if game.player == White then
+                Black
+
+            else
+                White
+    in
+    { game | player = newPlayer }
```
There is a slight risk that the student will start editing these definitions and function. One option to consider is to put this code in a second `editor` file (which are shown as read-only in the online editor, see [the docs](https://exercism.org/docs/building/tracks/concept-exercises#h-file-meta-config-json)).

This would also allow you to move the implementation of the rules from the tests file to the editor file.

ceddlyburge replied:
This is a good idea and I've done it, although the implementation of the rules is a test only thing, so I've left them where they are.

ErikSchierboom commented on exercises/concept/go/src/Go.elm:
```diff
@@ -0,0 +1,40 @@
+
+applyRules : Game -> Rule -> NonValidatingRule -> Rule -> Rule -> Game
+applyRules game oneStonePerPointRule captureRule libertyRule koRule =
+    Debug.todo "Please implement this function"
```
Tiny suggestion:

```suggestion
    Debug.todo "Please implement the `applyRules` function"
```

Or something similar. It might help a student in the online editor quickly identify what they need to do.

ErikSchierboom commented on exercises/concept/go/.meta/Exemplar.elm:
```diff
@@ -0,0 +1,53 @@
+    }
+
+
+type alias Rule =
+    Game -> Result String Game
```
This function type alias, is that something the student will be familiar with?

ceddlyburge replied:
Hi Erik, type alias is explained as part of records (to which it is tightly linked, which is a prerequisite of this xercise, so yes!

ErikSchierboom commented on concepts/result/introduction.md:
```diff
@@ -0,0 +1,16 @@
+This is known as a "custom type" definition in Elm terminology.
+Custom types are introduced in details in another concept of this track but we won't need it to understand specifically `Result`.
+The `error` and `value` in `OK value` and `Error error` represent a type variable, meaning it can be any type, such as `Int`, `Bool` or `String`.
+As such, a `Result String Int` is a variable that is expected to contain an `Int`, and if there is an error this will be represented as a `String`.
```
I think that it might be good to mention that there is a `Result` module with which the student can work with `Result` types, as the student will need that information to solve the exercise.

ceddlyburge replied:
Hi Erik, good idea, I've done this.

ErikSchierboom reviewed (commented):
Great to see more Concept Exercises! I've not yet read all files in detailed, as I first wanted to get out the more general comments.

ceddlyburge commented on concepts/result/introduction.md:
````diff
@@ -0,0 +1,16 @@
+```
+
+This is known as a "custom type" definition in Elm terminology.
+Custom types are introduced in details in another concept of this track but we won't need it to understand specifically `Result`.
````
Good point, I've pushed a commit

ceddlyburge commented:
Hi @ErikSchierboom , I think I've addressed all the comments, so if you could take another look that would be cool. The build script doesn't really cope with multiple files in the src directory (one of your suggestions), but I've made a hack temporary fix for now. Cheers, Cedd

ErikSchierboom commented on config.json:
```diff
@@ -211,6 +216,18 @@
+          "result"
+        ],
+        "prerequisites": [
+          "maybe"
```
Just to be sure: this does not yet list all prerequisites mentioned in the design document: https://github.com/exercism/elm/pull/438/files#diff-df035168d36af2aa6b326789fa41efffae7f77b2c07a54dd0c26502cbafacd8dR24

The website will ignore any prerequisites that are not taught by any concept exercise (or that don't exist), so you _could_ add it to this list I think. But you can also do that later of course.

ceddlyburge replied:
Good spot, I've added these to the config.
